1338 add slots infiltration 1 infiltration 2 infiltration notes

bmeluch opened the pull request:
Adding "infiltrations" as a multivalued slot encompassing the EMSL metadata headers "infiltration_1", "infiltration_2", and "infiltration_notes"

turbomam reviewed (changes requested):
thanks @bmeluch 

I have left some suggestions. I think it might be helpful to meet by Zoom.

Could you please show me some data, in its native form, that has these infiltration values?

bmeluch commented on src/schema/nmdc.yaml:
```diff
@@ -3039,6 +3040,21 @@ slots:
 
+  infiltrations:
+    annotations:
+      expected_value:
```
Ok, I'll take that out. 2 questions then - 
I see `annotations` in a bunch of existing slots, so do you mean it's just staying in for MIxS consistency and we're not using it in new slots?
Is `examples` fulfilling the same role as `annotations: expected_value` for our purposes?

bmeluch commented on src/schema/nmdc.yaml:
```diff
@@ -3039,6 +3040,21 @@ slots:
+        value: first infiltration time;second infiltration time;notes
+    description: The amount of time it takes to accomplish the first and second infiltration activities, and any details, issues, or context needed to understand the infiltration activity.
+    examples:
+      - value: 40 minutes;50 minutes;2017-08-24
```
Yes, Montana suggests hh:mm:ss to record typical time ranges

bmeluch commented on src/schema/nmdc.yaml:
```diff
@@ -3039,6 +3040,21 @@ slots:
+      - infiltration_notes
+    multivalued: true
+    list_elements_ordered: true
+    range: TextValue
```
Happy to use strings and not have to deal with the additional structure😄 
What would be the best way to make sure to write forward thinking PRs? Is it usually just a matter of discussing upcoming changes here in the comments, or is there a better place I should be looking for them?

mslarae13 replied:
Can we change this range to time, formatted as HH:MM:SS?

bmeluch replied:
The most recent commit includes the change - not sure how to make it show up here

bmeluch commented on src/schema/nmdc.yaml:
```diff
@@ -3039,6 +3040,21 @@ slots:
+      - infiltration_1
+      - infiltration_2
+      - infiltration_notes
+    multivalued: true
```
Discussed this with @mslarae13 today and we had some thoughts on changing the structure. 
First, it will be okay to leave out `infiltration_notes` - it's there in the EMSL metadata submission spreadsheet, but we couldn't find any examples of anyone actually filling it out. This means that all the infiltration information will now just be the same type (time durations) instead of duration+string
Second, there could be any number of infiltrations. So what we really want is a slot holding as many `hh:mm:ss` values as the user wants to enter. Which I think means using a multivalued slot as you originally suggested is still a good idea? https://github.com/microbiomedata/nmdc-schema/issues/1338

bmeluch replied:
as to the last part - I don't believe this is going to be suggested to MIxS

mslarae13 commented:
@turbomam i think Bea addressed your comments. Can you review this and let her know if it's ready to merge?

turbomam commented:
It looks like the structure is getting simpler. That's good.

I think we have some misunderstanding about what it means for a slot to be `multivalued` . The examples you have provided appear to be semicolon-delimited string representations of durations. Concatenating multiple values with a delimiter does not make the value multivalued.

If you can provide three valid example data files that show Biosamples with one, two and three infiltration values, I think it will be easier for us to understand one another. Please also provide at least one invalid example file. I have a feeling we'll need to assert a validation pattern.

turbomam commented:
Just discussed on Slack hangout:
- @bmeluch removes `multivalued` and `list_elements_ordered`
- @bmeluch modifies `description` or adds `comments` or `see_also` to provide background on what an infiltration is in this context
- @turbomam provides a `pattern`
  - we could use something like ^(?:(?:(?:[0-9]|[1-9][0-9]):[0-5]?[0-9]:[0-5]?[0-9])|(?:(?:[0-9]|[1-9][0-9]):[0-5]?[0-9])|(?:(?:[0-9]|[1-9][0-9])):[0-5]?[0-9])(?:;(?:(?:(?:[0-9]|[1-9][0-9]):[0-5]?[0-9]:[0-5]?[0-9])|(?:(?:[0-9]|[1-9][0-9]):[0-5]?[0-9])|(?:(?:[0-9]|[1-9][0-9])):[0-5]?[0-9]))*$
  - we can experiment with that at https://regexr.com/7ptlm 

Later we can update this slot definition here in nmdc-schema to specify that MongoDB should store the individual infiltration times in a list. At that point, the slot would be `multivalued`, and we'd have to add some logic to https://github.com/microbiomedata/submission-schema/blob/main/sheets_and_friends/tsv_in/modifications_long.tsv to locally use the original delimited style! 

We can review this with @pkalita-lbl. Maybe he can also tell us about the tools that convert the **data** from the delimited style to the list style.

mslarae13 commented:
> I think we have some misunderstanding about what it means for a slot to be multivalued . The examples you have provided appear to be semicolon-delimited string representations of durations. Concatenating multiple values with a delimiter does not make the value multivalued.

The ':' is not the delimiter. It's the format. Infiltration time 1 = HH:MM:SS ... so 00:01:04 and infiltration time 2 is 00:02:03

So the example for multivalued is

00:01:04 | 00:02:03 @bmeluch can you check and make sure this is how the example and slot is written.

> If you can provide three valid example data files that show Biosamples with one, two and three infiltration values, I think it will be easier for us to understand one another. Please also provide at least one invalid example file. I have a feeling we'll need to assert a validation pattern.

Should only ever be 1 or 2. @bmeluch can you add examples?

turbomam commented:
I'll add customization of that slot for the submission portal in a few minutes

turbomam commented:
Then you would add  a line like this to
https://github.com/microbiomedata/submission-schema/blob/main/sheets_and_friends/tsv_in/import_slots_regardless.tsv

| source class | source file or URL                                                                                              | slot          | section | column order | destination class |
|--------------|-----------------------------------------------------------------------------------------------------------------|---------------|---------|--------------|-------------------|
| Biosample    | https://raw.githubusercontent.com/microbiomedata/nmdc-schema/v?.?.?/nmdc_schema/nmdc_materialized_patterns.yaml | infiltrations | ?       | ?            | SoilInterface     |

And lines like this to 
https://github.com/microbiomedata/submission-schema/blob/main/sheets_and_friends/tsv_in/modifications_long.tsv

Don't forget to fill in the cells where I left question marks `?`

bmeluch commented:
> Here is a structured slot definition for infiltrations, which could be used in the nmdc-schema
I added this definition plus a see_also link to the MONet infiltration protocol for more information.

> Should only ever be 1 or 2. @bmeluch can you add examples?
I thought we said there could be any number of infiltrations in case it won't always be MONet samples (even if it's mostly used for MONet samples)? Regardless I added an example file with biosamples with 1, 2, and 3 infiltration values, and an invalid example file.

turbomam commented:
Your valid example files failed because I gave you a bad pattern. I have fixed it in-line above

`^(?:[0-9]|[1-9][0-9]|9[0-9]|0[0-9]|0[0-5][0-9]):[0-5][0-9]:[0-5][0-9]$`

bmeluch commented:
> Your valid example files failed because I gave you a bad pattern. I have fixed it in-line above
> 
> `^(?:[0-9]|[1-9][0-9]|9[0-9]|0[0-9]|0[0-5][0-9]):[0-5][0-9]:[0-5][0-9]$`

I checked this regex against my valid example values before committing and they all matched. The error in the test details is saying 
`m.replace_doi_field_with_award_dois_list_field({'id': 123, 'doi': {'has_raw_value': 'not-a-url'}})`
Has there been some build test added re: DOIs since I made this branch? or something?

turbomam commented:
I'm seeing this as the most recent error now:

> ```
> yaml.scanner.ScannerError: mapping values are not allowed here
>   in "src/data/invalid/Biosample-invalid-infiltrations.yaml", line 2, column 10
> ```

Probably a YAML formatting error. I can look for closely before 9 am PT on Monday.

turbomam commented:
I don't know anything about new DOI tests but will keep my out for that

bmeluch commented:
> Probably a YAML formatting error. I can look for closely before 9 am PT on Monday.

you were right! thank you!
